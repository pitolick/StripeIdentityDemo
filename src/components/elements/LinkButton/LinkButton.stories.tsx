import type { Meta, StoryObj } from '@storybook/react'
import { LinkButton } from './LinkButton'

const meta: Meta<typeof LinkButton> = {
  title: 'Elements/LinkButton',
  component: LinkButton,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof LinkButton>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
  },
}

/**
 * カラーバリエーション：黒（ベタ）
 */
export const ColorBlackSolid: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'black',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：赤（ベタ）
 */
export const ColorRedSolid: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'red',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：青（ベタ）
 */
export const ColorBlueSolid: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'blue',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：グレー（ベタ）
 */
export const ColorGraySolid: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'gray',
    buttonStyle: 'solid',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：黒（線）
 */
export const ColorBlackOutline: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'black',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：赤（線）
 */
export const ColorRedOutline: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'red',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：青（線）
 */
export const ColorBlueOutline: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'blue',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * カラーバリエーション：グレー（線）
 */
export const ColorGrayOutline: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    color: 'gray',
    buttonStyle: 'outline',
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * 子要素にDOMを含める
 */
export const DOM_Children: Story = {
  args: {
    children: (
      <span className="text-center text-sm">
        Link
        <br />
        Button
      </span>
    ),
    href: '#',
  },
}

/**
 * インラインボタン
 */
export const InlineButton: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    full: false,
  },
}

/**
 * インラインボタン（アイコンあり）
 */
export const InlineButton_Icon: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    full: false,
    iconLeft: 'arrow',
    iconRight: 'arrow',
  },
}

/**
 * アイコン（左側Arrow）
 */
export const IconLeftArrow: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    iconLeft: 'arrow',
  },
}

/**
 * アイコン（左側SVG）
 */
export const IconLeft: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    iconLeft: (
      <svg className="h-auto w-2.5 rotate-180" width="17" height="16" viewBox="0 0 17 16" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path
          fillRule="evenodd"
          clipRule="evenodd"
          d="M1 7C0.447715 7 -3.73832e-07 7.44771 -3.49691e-07 8C-3.2555e-07 8.55228 0.447715 9 1 9L13.0858 9L7.79289 14.2929C7.40237 14.6834 7.40237 15.3166 7.79289 15.7071C8.18342 16.0976 8.81658 16.0976 9.20711 15.7071L16.0303 8.88388C16.5185 8.39573 16.5185 7.60427 16.0303 7.11612L9.20711 0.292893C8.81658 -0.0976318 8.18342 -0.0976318 7.79289 0.292893C7.40237 0.683417 7.40237 1.31658 7.79289 1.70711L13.0858 7L1 7Z"
          fill="currentColor"
        />
      </svg>
    ),
  },
}

/**
 * アイコン（右側Arrow）
 */
export const IconRightArrow: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    iconRight: 'arrow',
  },
}

/**
 * アイコン（右側SVG）
 */
export const IconRight: Story = {
  args: {
    children: 'LinkButton',
    href: '#',
    iconRight: (
      <svg className="h-auto w-2.5" width="17" height="16" viewBox="0 0 17 16" fill="none" xmlns="http://www.w3.org/2000/svg">
        <path
          fillRule="evenodd"
          clipRule="evenodd"
          d="M1 7C0.447715 7 -3.73832e-07 7.44771 -3.49691e-07 8C-3.2555e-07 8.55228 0.447715 9 1 9L13.0858 9L7.79289 14.2929C7.40237 14.6834 7.40237 15.3166 7.79289 15.7071C8.18342 16.0976 8.81658 16.0976 9.20711 15.7071L16.0303 8.88388C16.5185 8.39573 16.5185 7.60427 16.0303 7.11612L9.20711 0.292893C8.81658 -0.0976318 8.18342 -0.0976318 7.79289 0.292893C7.40237 0.683417 7.40237 1.31658 7.79289 1.70711L13.0858 7L1 7Z"
          fill="currentColor"
        />
      </svg>
    ),
  },
}

/**
 * 別タブで開く
 */
export const TargetBlank: Story = {
  args: {
    children: 'Go To Google',
    href: 'https://www.google.com/',
    target: '_blank',
  },
}
