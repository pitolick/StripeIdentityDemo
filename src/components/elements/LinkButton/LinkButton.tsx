/**
 * ボタン風のリンク
 * @module LinkButton
 */
import React from 'react'
import Link, { LinkProps } from 'next/link'

/**
 * ボタン風のリンクコンポーネントのプロパティ
 * @typedef {Object} LinkButtonProps
 * @property {string} children - ボタンに表示するテキスト
 * @property {boolean} full - ブロックボタンにするか
 * @property {LinkProps} linkProps - Next.jsのLinkコンポーネントのプロパティ
 * @property {React.AnchorHTMLAttributes<HTMLAnchorElement>} - アンカータグのHTML属性（hrefが競合するためLinkPropsを優先）
 */
interface LinkButtonProps extends LinkProps, Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, 'href'> {
  /**
   * 子要素
   */
  children: React.ReactNode
  /**
   * 色（デフォルトカラーはコンポーネント側で指定）
   */
  color?: 'black' | 'red' | 'blue' | 'gray'
  /**
   * ボタンのスタイル
   */
  buttonStyle?: 'solid' | 'outline'
  /**
   * trueでブロック要素化
   */
  full?: boolean
  /**
   * アイコンを設置（左側）
   * `arrow`で組み込みの矢印を表示
   */
  iconLeft?: React.ReactNode | 'arrow'
  /**
   * アイコンを設置（右側）
   * `arrow`で組み込みの矢印を表示
   */
  iconRight?: React.ReactNode | 'arrow'
}

/**
 * ボタン風のリンクコンポーネント
 * @param {LinkButtonProps} props - ボタン風のリンクコンポーネントのプロパティ
 * @returns {React.FC} - ボタン風のリンクコンポーネント
 */
export const LinkButton: React.FC<LinkButtonProps> = ({ children, color = 'blue', buttonStyle = 'solid', full = true, iconLeft, iconRight, ...linkProps }) => {
  // 色
  let colorClass = ''
  let iconColorClass = ''
  switch (color) {
    case 'black':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-black hover:bg-gray-900'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-black text-black bg-white hover:text-white hover:bg-black'
          iconColorClass = 'text-black group-hover:text-white'
          break
      }
      break
    case 'red':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-red hover:bg-red-light'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-red text-red bg-white hover:text-white hover:bg-red'
          iconColorClass = 'text-red group-hover:text-white'
          break
      }
      break
    case 'blue':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-white bg-blue hover:bg-blue-light'
          iconColorClass = 'text-white'
          break
        case 'outline':
          colorClass = 'border-blue text-blue bg-white hover:text-white hover:bg-blue'
          iconColorClass = 'text-blue group-hover:text-white'
          break
      }
      break
    case 'gray':
      switch (buttonStyle) {
        case 'solid':
          colorClass = 'border-transparent text-gray-900 bg-gray-500 hover:bg-gray-300'
          iconColorClass = 'text-gray-900'
          break
        case 'outline':
          colorClass = 'border-gray-500 text-gray-500 bg-white hover:border-transparent hover:text-gray-900 hover:bg-gray-500'
          iconColorClass = 'text-gray-500 group-hover:text-gray-900'
          break
      }
      break
  }

  // アイコン用の共通クラス
  const iconBaseClass = 'absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5 '
  return (
    <Link
      className={`
        ${full && 'w-full'} ${colorClass} group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      `}
      {...linkProps}
    >
      {iconLeft && <span className={`${iconBaseClass} ${iconColorClass} left-4`}>{iconLeft}</span>}
      {children}
      {iconRight && <span className={`${iconBaseClass} ${iconColorClass} right-4`}>{iconRight}</span>}
    </Link>
  )
}
