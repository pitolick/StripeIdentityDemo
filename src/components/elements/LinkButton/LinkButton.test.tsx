import React from 'react'
import { render } from '@testing-library/react'
import { LinkButton } from './LinkButton'

describe('LinkButton', () => {
  it('正しくレンダリングされること', () => {
    const { container } = render(<LinkButton href={'test'}>テストボタン</LinkButton>)
    expect(container).toMatchSnapshot()
  })

  it('class名が正しく設定されること', () => {
    const { container } = render(
      <LinkButton href={'test'} className={'text-red'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが黒色でbuttonStyleがsolidの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'black'} buttonStyle={'solid'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが黒色でbuttonStyleがoutlineの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'black'} buttonStyle={'outline'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが赤色でbuttonStyleがsolidの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'red'} buttonStyle={'solid'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが赤色でbuttonStyleがoutlineの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'red'} buttonStyle={'outline'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが青色でbuttonStyleがsolidの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'blue'} buttonStyle={'solid'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが青色でbuttonStyleがoutlineの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'blue'} buttonStyle={'outline'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが灰色でbuttonStyleがsolidの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'gray'} buttonStyle={'solid'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('colorが灰色でbuttonStyleがoutlineの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} color={'gray'} buttonStyle={'outline'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('fullがfalseの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} full={false}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('fullがtrueの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} full={true}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('iconLeftがarrowの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} iconLeft={'arrow'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('iconLeftがarrow以外の場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} iconLeft={<div>test</div>}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('iconRightがarrowの場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} iconRight={'arrow'}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })

  it('iconRightがarrow以外の場合、正しくレンダリングされること', () => {
    const { container } = render(
      <LinkButton href={'test'} iconRight={<div>test</div>}>
        テストボタン
      </LinkButton>,
    )
    expect(container).toMatchSnapshot()
  })
})
