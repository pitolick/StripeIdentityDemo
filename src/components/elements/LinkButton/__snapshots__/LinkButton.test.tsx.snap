// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton class名が正しく設定されること 1`] = `
<div>
  <a
    class="text-red"
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが灰色でbuttonStyleがoutlineの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-gray-500 text-gray-500 bg-white hover:border-transparent hover:text-gray-900 hover:bg-gray-500 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが灰色でbuttonStyleがsolidの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-gray-900 bg-gray-500 hover:bg-gray-300 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが赤色でbuttonStyleがoutlineの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-red text-red bg-white hover:text-white hover:bg-red group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが赤色でbuttonStyleがsolidの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-red hover:bg-red-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが青色でbuttonStyleがoutlineの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-blue text-blue bg-white hover:text-white hover:bg-blue group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが青色でbuttonStyleがsolidの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが黒色でbuttonStyleがoutlineの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-black text-black bg-white hover:text-white hover:bg-black group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton colorが黒色でbuttonStyleがsolidの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-black hover:bg-gray-900 group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton fullがfalseの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        false border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton fullがtrueの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;

exports[`LinkButton iconLeftがarrowの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white left-4"
    >
      arrow
    </span>
    テストボタン
  </a>
</div>
`;

exports[`LinkButton iconLeftがarrow以外の場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white left-4"
    >
      <div>
        test
      </div>
    </span>
    テストボタン
  </a>
</div>
`;

exports[`LinkButton iconRightがarrowの場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white right-4"
    >
      arrow
    </span>
  </a>
</div>
`;

exports[`LinkButton iconRightがarrow以外の場合、正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
    <span
      class="absolute inset-y-0 inline-flex items-center justify-center m-auto h-full w-3.5  text-white right-4"
    >
      <div>
        test
      </div>
    </span>
  </a>
</div>
`;

exports[`LinkButton 正しくレンダリングされること 1`] = `
<div>
  <a
    class="
        w-full border-transparent text-white bg-blue hover:bg-blue-light group relative inline-flex items-center justify-center gap-2 rounded-md border-2 px-10 py-2.5 text-base  font-bold transition-all
      "
    href="test"
  >
    テストボタン
  </a>
</div>
`;
